### PR TITLE
xcode: allow ICMP ping relay on macOS + iOS/tvOS platforms

### PR DESCRIPTION
**Pull request by agottardo**
Fixes tailscale/tailscale#10393
Fixes tailscale/corp#15412
Fixes tailscale/corp#19808

On Apple platforms, exit nodes and subnet routers have been unable to relay pings from Tailscale devices to non-Tailscale devices due to sandbox restrictions imposed on our network extensions by Apple. The sandbox prevented the code in netstack.go from spawning the `ping` process which we were using.

Replace that exec call with logic to send an ICMP echo request directly, which appears to work in userspace, and not trigger a sandbox violation in the syslog.

### CODE REVIEW

File: wgengine/netstack/netstack.go [hunk 1]
```diff
@@ -891,7 +891,46 @@ func (ns *Impl) userPing(dstIP netip.Addr, pingResPkt []byte) {
...
+		ifState := ns.dialer.NetMon().InterfaceState()
+		defaultIfIPs := ifState.InterfaceIPs[ifState.DefaultRouteInterface]
+		var srcIP netip.Addr
+		for _, pfx := range defaultIfIPs {
+			addr := pfx.Addr()
+			if dstIP.Is4() && addr.Is4() || dstIP.Is6() && addr.Is6() {
+				srcIP = addr
+				break
+			}
+		}
```
Comment (bradfitz) [inline wgengine/netstack/netstack.go:906]: maybe pull this out to a method on the netmon.State? Like `SrcIPForDstIP(dstIP) (_ netip.Addr, ok bool)` ?

Then handle `!ok`, which currently you don't and fall through to making a packet with a zero header?

File: wgengine/netstack/netstack.go [hunk 2]
```diff
@@ -891,7 +891,46 @@ func (ns *Impl) userPing(dstIP netip.Addr, pingResPkt []byte) {
...
+					Src:     srcIP,
+					Dst:     dstIP,
+				},
+				Type: packet.ICMP4EchoRequest,
+				Code: packet.ICMP4NoCode,
+			}
+		}
+
+		icmpPing := packet.Generate(icmph, payload)
+		err = ns.tundev.InjectOutbound(icmpPing)
```
Comment (bradfitz) [inline wgengine/netstack/netstack.go:932]: you're sending it off but not waiting for the reply too?

Comment (agottardo) [reply wgengine/netstack/netstack.go]: Fixed

File: wgengine/netstack/netstack.go (hunk 1 above)
Comment (agottardo) [reply wgengine/netstack/netstack.go]: Went with a different approach (try to open a connection and check what the srcIP there is)

File: wgengine/netstack/netstack_userping_apple.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build darwin || ios
+
+package netstack
+
+import (
+	"github.com/prometheus-community/pro-bing"
```
Comment (bradfitz) [inline wgengine/netstack/netstack_userping_apple.go:9]: no goimports formatting on this? normally non-std imports go in their own section after a blank line

so:

```go
import (
   "net/netip"
    "time"

    "github.com/...."
)
```

**Review by bradfitz — APPROVED** [review]

File: wgengine/netstack/netstack_userping_apple.go (hunk 3 above)
Comment (agottardo) [reply wgengine/netstack/netstack_userping_apple.go]: fixed.

**Pull request merged**